src/main/java/hudson/plugins/ec2/EC2Cloud.java

```diff
@@ -1189,10 +1189,11 @@ public Node call() throws Exception {
 
                             InstanceStateName state = instance.state().name();
                             if (state.equals(InstanceStateName.RUNNING)) {
-                                // Spot instance are not reconnected automatically,
-                                // but could be new orphans that has the option enable
+                                // JENKINS-76171: Always initiate connection when instance is RUNNING
+                                // This reduces the gap between PlannedNode completion and agent connection,
+                                // preventing over-provisioning in NodeProvisioner
                                 Computer c = slave.toComputer();
-                                if (slave.getStopOnTerminate() && (c != null)) {
+                                if (c != null) {
```

Comment:
IIUC this is the primary fix.

Reply:
no, that helps but not under conditions where you are really creating many jobs rapidly.

https://github.com/jenkinsci/ec2-plugin/pull/1149/files/c457ff9aa25b7a400d5927a2ef17fcf8f33b275d#diff-f917fcc19904db51b41bee342f6b90ab9cb00a853343b84dff70ffd116951590R162-R171

and adding the `synchronized` where what really fixed the issue